pkg/test/server.go

```diff
@@ -0,0 +1,125 @@
+package test
+
+import (
+	"encoding/csv"
+	"fmt"
+	"net"
+	"os"
+
+	"github.com/miekg/dns"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sync/errgroup"
+)
+
+var domainToIP = loadRecords("testdata/test-300.csv")
+
+const truncateSize = 512
+
+type handler struct {
+	truncate bool
+}
+
+type Server struct {
+	Addr    string
+	TCPPort string
+	UDPPort string
+}
+
+func (s *Server) Run() {
+	g := new(errgroup.Group)
+	udpSrv := &dns.Server{Addr: net.JoinHostPort(s.Addr, s.UDPPort), Net: "udp", Handler: &handler{truncate: true}}
+	defer func() {
+		if err := udpSrv.Shutdown(); err != nil {
+			logrus.Errorf("shutting down UDP server: %v", err)
+		}
+	}()
+	g.Go(udpSrv.ListenAndServe)
+
+	tcpSrv := &dns.Server{Addr: net.JoinHostPort(s.Addr, s.TCPPort), Net: "tcp", Handler: &handler{}}
+	defer func() {
+		if err := tcpSrv.Shutdown(); err != nil {
+			logrus.Errorf("shutting down TCP server: %v", err)
+		}
+	}()
+	g.Go(tcpSrv.ListenAndServe)
+
+	if err := g.Wait(); err != nil {
+		logrus.Fatalf("ListenAndServe error: %v", err)
+	}
+}
+
+func (h *handler) ServeDNS(w dns.ResponseWriter, r *dns.Msg) {
+	msg := new(dns.Msg)
+	msg.SetReply(r)
+	msg.Compress = false
+	// we might handle other OpCodes later
+	if r.Opcode != dns.OpcodeQuery {
+		return
+	}
+	parseReply(msg)
+	if h.truncate {
+		msg.Truncate(truncateSize)
+	}
+	err := w.WriteMsg(msg)
+	if err != nil {
+		logrus.Errorf("ServeDNS trying to write message in udpHandler: %v", err)
+	}
+}
+
+func parseReply(msg *dns.Msg) {
+	logrus.Infof("ServeDNS handling request: %v", msg.Question)
```

Comment:
Do we expect to hammer this server a lot? If yes, `Info` may be too high a level for logging.

Reply:
This is not going to act as a recursive server, this will be the final upstream. There is no internet and we are planning to tap into win32 DNS API since it won't work without internet.